File: schemas/subscription-membership/drizzle/coupon_redemptions.ts
// coupon_redemptions: Tracks each use of a coupon by a customer.
// See README.md for full design rationale.

import {
  pgTable,
  uuid,
  text,
  timestamp,
  index,
} from "drizzle-orm/pg-core";
import { coupons } from "./coupons";
import { customers } from "./customers";
import { subscriptions } from "./subscriptions";

export const couponRedemptions = pgTable(
  "coupon_redemptions",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    couponId: uuid("coupon_id")
      .notNull()
      .references(() => coupons.id, { onDelete: "cascade" }),
    customerId: uuid("customer_id")
      .notNull()
      .references(() => customers.id, { onDelete: "cascade" }),
    subscriptionId: uuid("subscription_id").references(() => subscriptions.id, { onDelete: "set null" }),
    redeemedAt: timestamp("redeemed_at", { withTimezone: true }).notNull().defaultNow(),
    providerType: text("provider_type"),
    providerId: text("provider_id"),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    index("idx_coupon_redemptions_coupon_id").on(table.couponId),
    index("idx_coupon_redemptions_customer_id").on(table.customerId),
    index("idx_coupon_redemptions_subscription_id").on(table.subscriptionId),
  ]
);

File: schemas/subscription-membership/spacetimedb/coupon_redemptions.rs
// coupon_redemptions: Records of coupons applied to customer subscriptions.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[spacetimedb::table(name = coupon_redemptions, public)]
pub struct CouponRedemption {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub coupon_id: String, // UUID — FK → coupons.id (cascade delete)
    #[index(btree)]
    pub customer_id: String, // UUID — FK → customers.id (cascade delete)
    pub subscription_id: Option<String>, // UUID — FK → subscriptions.id (set null)
    pub redeemed_at: Timestamp,
    pub provider_type: Option<String>,
    pub provider_id: Option<String>,
    pub created_at: Timestamp,
}
